xds/src/main/java/io/grpc/xds/ServerXdsClient.java

```diff
@@ -101,15 +101,21 @@ public void run() {
   @Override
   Collection<String> getSubscribedResources(ResourceType type) {
     if (newServerApi) {
-      String listeningAddress = instanceIp + ":" + listenerPort;
-      String resourceName =
-          grpcServerResourceId + "?udpa.resource.listening_address=" + listeningAddress;
-      return ImmutableList.<String>of(resourceName);
+      return ImmutableList.<String>of(getResourceNameForListener());
     } else {
       return Collections.emptyList();
     }
   }
 
+  private String getResourceNameForListener() {
```

Comment:
This becomes really messy. You could have fix the resource name in the constructor instead of keeping recomputing it. That's much much cleaner.

Reply:
The `port` is not known in the constrictor so this can be built in `watchListenerData` which is where I compute it.